refactor(*): use bundled homs for `lift` in many `free_*` constructions

urkud opened the pull request:
This is a work in progress, help is welcome.

urkud committed:
chore(algebra/free_monoid): use implicit args in `lift`

urkud committed:
Snapshot

urkud committed:
free_ring: fix compile

urkud committed:
Snapshot

gebner commented on src/algebra/free_monoid.lean:
```diff
@@ -14,7 +14,7 @@ import data.list.basic
 * `free_monoid.of`: embedding `α → free_monoid α` sending each element `x` to `[x]`;
-* `free_monoid.lift α M`: natural equivalence between `α → M` and `free_monoid α →* M`;
+* `free_monoid.lift`: natural equivalence between `α → M` and `free_monoid α →* M`;
   for technical reasons `α` and `M` are explicit arguments;
```
```suggestion
* `free_monoid.lift`: natural equivalence between `α → M` and `free_monoid α →* M`
```

urkud committed:
Update src/algebra/free_monoid.lean

Co-authored-by: Gabriel Ebner <gebner@gebner.org>

urkud committed:
Merge branch 'master' into free-group-lift

urkud committed:
feat(group_theory/free_comm_monoid): define `free_comm_monoid`

Define `free_comm_monoid` as a simple wrapper over
`multiplicative (multiset α)`.

urkud committed:
Add a missing docstring

urkud committed:
Merge branch 'free-comm-monoid' into free-group-lift

urkud committed:
Merge branch 'master' into free-group-lift

urkud committed:
fix

urkud committed:
Merge branch 'master' into free-group-lift

urkud committed:
Snapshot

urkud committed:
Merge branch 'master' into free-group-lift
